File: apps/mobile/iosApp/Piro/Profile/ProfileViewModel.swift
import Foundation
import Shared

/// Loads and edits the signed-in user's profile: display name and avatar color (persisted via
/// `PUT /api/v1/auth/me`), plus their notification-delivery preferences (read-only for now).
@MainActor
final class ProfileViewModel: ObservableObject {
    @Published private(set) var profile: UserProfile?
    @Published private(set) var preferences: [UserNotificationPreferenceDto] = []
    @Published var loading = true
    @Published var error: String?

    // Editable fields, seeded from the loaded profile.
    @Published var name = ""
    @Published var colorHex = ""
    @Published var saving = false

    private let api: PiroApiClient

    init(api: PiroApiClient) {
        self.api = api
    }

    func load() async {
        loading = true
        error = nil
        do {
            let profile = try await api.me()
            apply(profile)
            // Preferences are best-effort — a user with none, or a permissions gap, shouldn't blank the page.
            preferences = (try? await api.getNotificationPreferences(userId: profile.id)) ?? []
        } catch {
            self.error = "Could not load your profile."
        }
        loading = false
    }

    var isDirty: Bool {
        guard let profile else { return false }
        return name != profile.name || colorHex != profile.color
    }

    func save() async {
        guard let profile, isDirty, !saving else { return }
        saving = true
        error = nil
        do {
            let updated = try await api.updateProfile(name: name, color: colorHex, timeZone: profile.timeZone)
            apply(updated)
        } catch {
            self.error = PiroError.message(error, networkFallback: "Could not save your changes.")
        }
        saving = false
    }

    private func apply(_ profile: UserProfile) {
        self.profile = profile
        name = profile.name
        colorHex = profile.color
    }
}

File: apps/mobile/androidApp/src/main/kotlin/co/heva/piro/android/profile/ProfileViewModel.kt
package co.heva.piro.android.profile

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import co.heva.piro.shared.api.PiroApiClient
import co.heva.piro.shared.model.UserProfile
import co.heva.piro.shared.model.generated.UserNotificationPreferenceDto
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class ProfileUiState(
    val loading: Boolean = true,
    val profile: UserProfile? = null,
    val preferences: List<UserNotificationPreferenceDto> = emptyList(),
    val error: String? = null,
)

/** Loads the signed-in user's profile and notification preferences for the Profile tab. */
class ProfileViewModel(private val api: PiroApiClient) : ViewModel() {

    private val _state = MutableStateFlow(ProfileUiState())
    val state: StateFlow<ProfileUiState> = _state.asStateFlow()

    init {
        load()
    }

    private fun load() {
        viewModelScope.launch {
            _state.update { it.copy(loading = true, error = null) }
            try {
                val profile = api.me()
                // Preferences are best-effort — a failure there shouldn't blank the whole profile.
                val prefs = runCatching { api.getNotificationPreferences(profile.id) }.getOrDefault(emptyList())
                _state.update { it.copy(loading = false, profile = profile, preferences = prefs) }
            } catch (e: Exception) {
                _state.update { it.copy(loading = false, error = "Could not load your profile.") }
            }
        }
    }
}
